File: business_logic/src/test/java/org/metplus/curriculum/domain/job/CreateJobTest.java
package org.metplus.curriculum.domain.job;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.Before;
import org.junit.Test;
import org.metplus.curriculum.domain.DomainFactories;
import org.metplus.curriculum.domain.Job;

public class CreateJobTest {
    private JobRepositoryFake jobRepositoryFake;
    private CreateJob<Void> createJob;
    @Before
    public void setup() {
        jobRepositoryFake = new JobRepositoryFake();
        createJob = new CreateJob<>(jobRepositoryFake);
    }

    @Test
    public void when_job_id_exists_it_calls_job_exists_callback() {
        Job job = DomainFactories.buildJob();
        job = jobRepositoryFake.save(job);
        CreateJobResultHandlerSpy resultHandler = new CreateJobResultHandlerSpy();
        String title = "";
        String description = "";
        createJob.create(job.getJobId(), title, description, resultHandler);
        assertThat(resultHandler.jobExistsWasCalledWith).isEqualTo(job.getJobId());
    }

    @Test
    public void when_error_happens_while_saving_it_calls_fatal_error_callback() {
        Job job = DomainFactories.buildJob();
        job = jobRepositoryFake.save(job);
        CreateJobResultHandlerSpy resultHandler = new CreateJobResultHandlerSpy();
        String title = "";
        String description = "";
        createJob.create(job.getJobId(), title, description, resultHandler);
        assertThat(resultHandler.fatalErrorWasCalled).isEqualTo(job.getJobId());
    }

    @Test
    public void when_job_id_do_not_exist_it_saves_in_database_and_call_success_callback() {
        Job job = DomainFactories.buildJob();
        job.setJobId("someJobId");
        CreateJobResultHandlerSpy resultHandler = new CreateJobResultHandlerSpy();
        createJob.create(job.getJobId(), job.getTitle(), job.getDescription(), resultHandler);

        Job createdJob = jobRepositoryFake.findById("someJobId");
        assertThat(job).isEqualToComparingFieldByField(createdJob);

        assertThat(resultHandler.successWasCalled).isEqualTo(true);
    }
}

File: business_logic/src/test/kotlin/org/metplus/curriculum/domain/job/CreateJobTest.kt
package org.metplus.curriculum.domain.job

import org.assertj.core.api.Assertions.assertThat

import org.junit.Before
import org.junit.Test
import org.metplus.curriculum.domain.DomainFactories
import org.metplus.curriculum.domain.job.testDoubles.CreateJobResultHandlerSpy
import org.metplus.curriculum.domain.job.testDoubles.JobRepositoryFake

class CreateJobTest {
    private lateinit var jobRepositoryFake: JobRepositoryFake
    private lateinit var createJob: CreateJob
    @Before
    fun setup() {
        jobRepositoryFake = JobRepositoryFake()
        createJob = CreateJob(jobRepositoryFake)
    }

    @Test
    fun `when jobId exists, it calls job exists callback`() {
        var job = DomainFactories.buildJob()
        job = jobRepositoryFake.save(job)
        val resultHandler = CreateJobResultHandlerSpy()
        val title = ""
        val description = ""
        createJob.create(
                jobId = job.jobId,
                title = title,
                description = description,
                resultHandler = resultHandler)
        assertThat(resultHandler.jobExistsWasCalledWith).isEqualTo(job.jobId)
    }

    @Test
    fun `when error happens while saving, it calls fatal error callback`() {
        jobRepositoryFake.saveThrowsException = Exception()
        val resultHandler = CreateJobResultHandlerSpy()
        val title = ""
        val description = ""
        createJob.create(
                jobId = "asd",
                title = title,
                description = description,
                resultHandler = resultHandler
        )
        assertThat(resultHandler.fatalErrorWasCalledWith).isEqualTo("asd")
    }

    @Test
    fun `when jobId do not exist, it saves in database and call success callback`() {
        val job = DomainFactories.buildJob()
        job.jobId = "someJobId"
        val resultHandler = CreateJobResultHandlerSpy()
        createJob.create(
                jobId = job.jobId,
                title = job.title,
                description = job.description,
                resultHandler = resultHandler
        )

        val createdJob = jobRepositoryFake.findById("someJobId")
        assertThat(job).isEqualToComparingFieldByField(createdJob)

        assertThat(resultHandler.successWasCalled).isEqualTo(true)
    }
}